vm/virtualbox/virtualbox.go

```diff
@@ -0,0 +1,312 @@
+// Copyright 2025 syzkaller project authors. All rights reserved.
+// Use of this source code is governed by Apache 2 LICENSE that can be found in the LICENSE file.
+
+package virtualbox
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"net"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strconv"
+	"time"
+
+	"github.com/google/syzkaller/pkg/config"
+	"github.com/google/syzkaller/pkg/log"
+	"github.com/google/syzkaller/pkg/osutil"
+	"github.com/google/syzkaller/pkg/report"
+	"github.com/google/syzkaller/sys/targets"
+	"github.com/google/syzkaller/vm/vmimpl"
+)
+
+func init() {
+	vmimpl.Register("virtualbox", vmimpl.Type{Ctor: ctor})
+}
+
+type Config struct {
+	BaseVM string `json:"vm_name"` // name of the base VM
+	Count  int    `json:"count"`   // number of VMs to run in parallel
+}
+
+type Pool struct {
+	env *vmimpl.Env
+	cfg *Config
+}
+
+type instance struct {
+	cfg        *Config
+	debug      bool
+	baseVM     string
+	vmName     string
+	sshPort    int
+	rpcPort    int
+	sshuser    string
+	sshkey     string
+	timeouts   targets.Timeouts
+	serialPath string
+	closed     chan bool
+}
+
+func ctor(env *vmimpl.Env) (vmimpl.Pool, error) {
+	cfg := &Config{}
+	if err := config.LoadData(env.Config, cfg); err != nil {
+		return nil, err
+	}
+	if cfg.BaseVM == "" {
+		return nil, fmt.Errorf("config param base_vm is empty")
+	}
+	if cfg.Count < 1 || cfg.Count > 128 {
+		return nil, fmt.Errorf("invalid config param count: %v, want [1,128]", cfg.Count)
+	}
+	if _, err := exec.LookPath("VBoxManage"); err != nil {
+		return nil, fmt.Errorf("cannot find VBoxManage")
+	}
+	return &Pool{cfg: cfg, env: env}, nil
+}
+
+func (pool *Pool) Count() int { return pool.cfg.Count }
+
+func (pool *Pool) Create(_ context.Context, workdir string, index int) (vmimpl.Instance, error) {
+	timestamp := strconv.FormatInt(time.Now().UnixNano(), 10)
+	serialPath := filepath.Join(workdir, timestamp, "serial")
```

Comment:
You can simplify this code by dropping timestamp - workdir is already unique for this particular VM instance

https://github.com/google/syzkaller/blob/4e1406b4defac0e2a9d9424c70706f79a7750cf3/vm/vm.go#L164-L167
https://github.com/google/syzkaller/blob/4e1406b4defac0e2a9d9424c70706f79a7750cf3/vm/vm.go#L173-L177